crates/dbsp/src/operator/dynamic/balance/balancer.rs

```diff
@@ -0,0 +1,833 @@
+        maxsat::{Cost, HardConstraint, Variable, VariableIndex},
+        MaxSat,
+    },
+    utils::components,
+    Error, Runtime,
+};
+use num::abs;
+use petgraph::graph::UnGraph;
+use serde::{Deserialize, Serialize};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet},
+    fmt::{Display, Error as FmtError, Formatter},
+    rc::Rc,
+};
+
+const MIN_RELATIVE_IMPROVEMENT_THRESHOLD: f64 = 1.5;
+const MIN_ABSOLUTE_IMPROVEMENT_THRESHOLD: Cost = Cost(0);
+
+#[derive(Debug, Clone, Serialize)]
+pub enum BalancerError {
+    /// We currently only support balancing top-level nodes.
+    NonTopLevelNode(GlobalNodeId),
+    /// Trying to set a hint for a stream that is not registered with the balancer.
+    NotRegisteredWithBalancer(NodeId),
+    /// Hint cannot be enforced in the current state.
+    InvalidPolicyHint(PartitioningPolicy, String),
+    /// No solution found for .
+    NoSolution,
+}
+
+impl Display for BalancerError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self {
+            Self::NonTopLevelNode(node) => write!(f, "node {node} is not a top level node"),
+            Self::NotRegisteredWithBalancer(node) => {
+                write!(f, "node {node} is not registered with the balancer")
+            }
+            // TODO: this is the only error that should normally propagate to the user.
+            // (others are Feldera bugs). We may need a better error description here.
+            Self::InvalidPolicyHint(policy, message) => {
+                write!(
+                    f,
+                    "policy hint {policy:?} cannot be enforced in the current state: {message}"
+                )
+            }
+            Self::NoSolution => f.write_str("no solution found for balancing constraints"),
+        }
+    }
+}
+
+/// A MaxSat constraint that enforces that the two streams are partitioned
+/// using one of the following compatible policies (if `is_left_join` is false):
+///
+/// ```text
+/// ┌─────────---──┬────────---------───┐
+/// │    Left      │      Right         │
+/// ├────────---───┼─────---------──────┤
+/// │ Shard        │ Shard or Broadcast │
+/// ├────────---───┼────---------───────┤
+/// │ Broadcast    │ Shard or Balance   │
+/// ├────────---───┼────---------───────┤
+/// │ Balance      │ Broadcast          │
+/// └────────---───┴────---------───────┘
+/// ```
+///
+/// or, if `is_left_join` is true:
+///
+/// ```text
+/// ┌─────────---──┬────────---------───┐
+/// │    Left      │      Right         │
+/// ├────────---───┼─────---------──────┤
+/// │ Shard        │ Shard or Broadcast │
+/// ├────────---───┼────---------───────┤
+/// │ Balance      │ Broadcast          │
+/// └────────---───┴────---------───────┘
+/// ```
+
+#[derive(Debug, Clone)]
+pub struct JoinConstraint {
+    v1: VariableIndex,
+    v2: VariableIndex,
+    is_left_join: bool,
+}
+
+impl JoinConstraint {
+    pub fn create_join_constraint(
```

Comment:
the name suggests that the constraint is returned
It may be cleaner to separate constraint creation from the solver API